include/aos/common/tools/map.hpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2023 Renesas Electronics Corporation.
+ * Copyright (C) 2023 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef AOS_MAP_HPP_
+#define AOS_MAP_HPP_
+
+#include "aos/common/tools/array.hpp"
+#include "aos/common/tools/utils.hpp"
+
+namespace aos {
+
+/**
+ * Simple map implementation based on a non sorted array.
+ */
+template<typename Key, typename Value>
+class Map {
+public:
+    /**
+     * Returns pointer to the value for a given key. And null if a key is absent.
+     * 
+     * @param key
+     * @return const Value*
+     */
+    Value* Get(const Key& key);
+
+    /**
+     * Returns pointer to the value for a given key. And null if a key is absent.
+     * 
+     * @param key
+     * @return const Value*
+     */
+    const Value* Get(const Key& key) const;
+
+    /**
+     * Inserts or replaces a value with a specified key.
+     * 
+     * @param key
+     * @param value
+     * @return Error
+     */
+    Error Set(const Key& key, const Value& value);
+
+    /**
+     * Removes element with the specified key from the map.
+     * 
+     * @param key
+     * @return Error
+     */
+    Error Erase(const Key& key);
+
+    /**
+     * Removes all elements from the map.
+     */
+    void Clear();
+
+    /**
+     * Returns current number of elements in the map.
+     *
+     * @return size_t.
+     */
+    size_t Size() const;
+
+    /**
+     * Returns maximum  allowed number of elements in the map.
+     *
+     * @return size_t.
+     */
+    size_t Capacity() const;
+
+    /**
+     * Iterators to the beginning / end of the map 
+     */
+    Pair<Key,Value>*       begin();
+    Pair<Key,Value>*       end();
+    const Pair<Key,Value>* begin() const;
+    const Pair<Key,Value>* end() const;
+
+protected:
+    Map(Array<Pair<Key,Value>>& array);
+};
+
+}
+
```

Comment:
What about operator[] ?

Reply:
operator[] inserts key if it is absent & for our case should return  RetWithError<Value&> or Value*.  This makes its usage different from builtin operators, that looks weird.
If required I can add Insert function which behaves the same.